drafts/builder-pattern-tables.md

```diff
@@ -0,0 +1,354 @@
+
+```typescript
+/**
+* Builder that returns a column configuration to be used by the material table.
+*/
+export class TableColumnConfigBuilder<T> {
+  private columnDef: TableColumnConfiguration<T>;
+
+  constructor() {
+    this.columnDef = this.defaultColumn();
+  }
+
+  /**
+   * Sets the column header and column id. The id is used internally by angular material and has to be unique within
+   * the table.
+   * @param title - header title
+   * @param id - unique column id e.g. fromDate
+   */
+  column(title: string, id: string): TableColumnConfigBuilder<T> {
+    this.columnDef = {
+      ...this.defaultColumn(),
+      title: title,
+      columnId: id,
+    };
+    return this;
+  }
+
+  /**
+   * Sets the column content type to a date like type. That means that we set the formatting function to a date formatter
+   * and the column min and max width automatically.
+   */
+  date(
+    valueGetterFn: (data: T) => TableCellContentValue,
+    format = 'mediumDate'
+  ) {
+    this.columnDef = {
+      ...this.columnDef,
+      cellContent: {
+        ...this.columnDef.cellContent,
+        valueGetterFn: valueGetterFn,
+        formatFn: (value: TableCellContentTypes) => {
+          if (typeof value === 'string') {
+            return esasDate(value, format);
+          }
+          return value;
+        },
+      },
+      styling: {
+        ...this.columnDef.styling,
+        minWidth: '105px',
+        maxWidth: '180px',
+      },
+    };
+    return this;
+  }
+
+  /**
+   * Sets the min and max width of the column to accommodate a typical PRIMA ID or user ID
+   */
+  id(valueGetterFn: (data: T) => TableCellContentValue) {
+    this.columnDef = {
+      ...this.columnDef,
+      cellContent: {
+        ...this.columnDef.cellContent,
+        displayType: TableCellDisplayingType.Value,
+        valueGetterFn: valueGetterFn,
+      },
+      styling: {
+        ...this.columnDef.styling,
+        minWidth: '80px',
+        maxWidth: '165px',
+      },
+    };
+    return this;
+  }
+
+  ...
+
+  /**
+   * Build step returning the column configuration as defined using the builder
+   */
+  build(): TableColumnConfiguration<T> {
+    return this.columnDef;
+  }
+}
+```
+
+The beneficial part of this builder pattern is that we can easily modify all tables in our application at once. If we e.g. decide to change the formatting of our date columns, we can easily change it in the builder, and it will be effective for all date columns in our application. Additionally, we have consistency by design, so an ID column is always the same size, or a multi-line list is always displayed the same.
+
+Moreover, the template of the components containing tables is clean and not clustered with table boilerplate code anymore. It just requires the two inputs, and thats it. Furthermore, it can be tested quite easily with unit and component tests.
```

Comment:
```suggestion
Moreover, the template of the components containing tables is clean and not clustered with table boilerplate code anymore. It just requires the two inputs, and that's it. Furthermore, it can be tested quite easily with unit and component tests.
```